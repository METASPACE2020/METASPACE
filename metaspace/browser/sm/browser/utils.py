import pathlib
import shutil


def list_file_sizes(bucket):
    obj_sizes = {}
    i = 0
    for obj in bucket.objects.all():
        obj_sizes[obj.key] = obj.size
        i += 1
        if i % 1000:
            print(f"{i} objects")

    obj_sizes = sorted(obj_sizes.items(), key=lambda kv: kv[1], reverse=True)

    obj_sizes_5gbmax = [(key, size) for key, size in obj_sizes if size < 5 * 1024 ** 3]
    for i in range(10):
        key, size = obj_sizes_5gbmax[i]
        print(size / 1024 ** 2, key)


def download_dataset(bucket, key_prefix, ds_local_path):
    ds_local_path.mkdir(parents=True, exist_ok=True)
    for obj in bucket.objects.filter(Prefix=key_prefix):
        # print(obj)
        fn = obj.key.split('/')[-1]
        f_path = ds_local_path / fn
        if not f_path.exists():
            bucket.download_file(obj.key, str(f_path))


def get_file_by_ext(path, ext):
    return [f_path for f_path in path.iterdir() if str(f_path).lower().endswith(ext)][0]


def clean_dir(path):
    shutil.rmtree(path, ignore_errors=True)
    path.mkdir(parents=True, exist_ok=True)


def mz_ppm_bin(mz, ppm):
    return mz - mz * ppm * 1e-6, mz + mz * ppm * 1e-6
