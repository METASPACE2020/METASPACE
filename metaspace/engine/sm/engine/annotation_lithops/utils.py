from __future__ import annotations

import json
import logging
from base64 import urlsafe_b64encode
from datetime import datetime
from hashlib import blake2b
from pathlib import Path

import numpy as np
import pandas as pd

# logging.getLogger('ibm_boto3').setLevel(logging.WARNING)
# logging.getLogger('ibm_botocore').setLevel(logging.WARNING)
# logging.getLogger('urllib3').setLevel(logging.WARNING)

logger = logging.getLogger('annotation-pipeline')


def ds_dims(coordinates):
    min_x, min_y = np.amin(coordinates, axis=0)[:2]
    max_x, max_y = np.amax(coordinates, axis=0)[:2]
    nrows, ncols = max_y - min_y + 1, max_x - min_x + 1
    return nrows, ncols


def get_pixel_indices(coordinates):
    _coord = np.array(coordinates)[:, :2]
    _coord = np.around(_coord, 5)
    _coord -= np.amin(_coord, axis=0)

    _, ncols = ds_dims(coordinates)
    pixel_indices = _coord[:, 1] * ncols + _coord[:, 0]
    pixel_indices = pixel_indices.astype(np.int32)
    return pixel_indices


def jsonhash(obj) -> str:
    """
    Calculates a hash for a JSON-stringifiable object. Intended for compacting large sets of
    parameters into a simple key that can be used to distinctly identify a cache entry.

    The output is collision-resistant, but shouldn't be assumed to be cryptographically secure.
    In most cases a motivated adversary could figure out the original object contents easily, as
    there's no hidden key and it's unlikely there will be much variety in the objects hashed.
    """
    json_val = json.dumps(obj, sort_keys=True)
    hash_val = blake2b(json_val.encode(), digest_size=12).digest()
    return str(urlsafe_b64encode(hash_val), 'utf-8')
