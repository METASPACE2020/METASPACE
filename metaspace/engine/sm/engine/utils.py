import logging
import os
import sys
from shutil import rmtree
import numpy as np
import pandas as pd

from pyspark.conf import SparkConf
from pyspark.context import SparkContext
from pyspark.sql.session import SparkSession

from sm.engine.db import DB
from sm.engine.formula_centroids import CentroidsGenerator
from sm.engine.isocalc_wrapper import IsocalcWrapper
from sm.engine.mol_db import MolDBServiceWrapper, MolecularDB
from sm.engine.msm_basic.formula_imager_segm import gen_iso_images
from sm.engine.msm_basic.formula_img_validator import formula_image_metrics, make_compute_image_metrics
from sm.engine.util import init_loggers

init_loggers()
logger = logging.getLogger('annotate-daemon')


def create_spark_context(sc=None, cores='*'):
    os.environ.setdefault('PYSPARK_PYTHON', sys.executable)

    sconf = SparkConf()
    sconf.set('spark.sql.execution.arrow.enabled', True)
    sconf.set('spark.executor.memory', '2g')
    sconf.set('spark.driver.memory', '4g')
    sconf.set("spark.driver.maxResultSize", "3g")
    sconf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    sconf.set("spark.kryoserializer.buffer.max", "512m")
    sconf.set("spark.executor.pyspark.memory", "1g")
    sconf.set("spark.python.worker.memory", "1g")
    sconf.set("spark.rdd.compress", True)

    if sc:
        sc.stop()
    sc = SparkContext(master=f'local[{cores}]', appName='sm engine notebook', conf=sconf)
    spark_session = SparkSession(sc)
    return sc, spark_session


def fetch_moldbs(sm_config, iso_gen_config):
    db = DB(sm_config['db'])
    moldb_service = MolDBServiceWrapper(sm_config['services']['mol_db'])
    moldb_id = moldb_service.find_db_by_name_version('HMDB-v4')[0]['id']
    moldbs = [MolecularDB(id=moldb_id, db=db, iso_gen_config=iso_gen_config)]
    return moldbs


def gen_formula_centroids(sc, iso_gen_config, ion_formulas):
    isocalc = IsocalcWrapper(iso_gen_config)
    centroids_gen = CentroidsGenerator(sc=sc, isocalc=isocalc)
    formula_centroids = centroids_gen.generate_if_not_exist(formulas=ion_formulas.tolist())
    # centroids_df = formula_centroids.centroids_df()
    return formula_centroids


def ds_sample_gen(imzml_parser, sp_n, sample_ratio=0.05, max_sample_size=1000):
    sample_size = min(max_sample_size, int(sp_n * sample_ratio))
    sample_sp_inds = np.random.choice(np.arange(sp_n), sample_size)
    for sp_idx in sample_sp_inds:
        mzs, ints = imzml_parser.getspectrum(sp_idx)
        yield sp_idx, mzs, ints


def define_mz_segments(imzml_parser, centroids_df, sample_ratio=0.05, mz_overlap=8, ppm=3):

    def optimal_segm_n(total_n_mz, mz_per_segm, min_i, max_i):
        n = round(total_n_mz / mz_per_segm)
        i = np.argmin([abs(n - 2 ** i) for i in range(min_i, max_i)])
        return 2 ** (min_i + i)

    def bounds_to_segments(segm_bounds):
        mz_segments = []
        for i, (l, r) in enumerate(zip(segm_bounds[:-1],
                                       segm_bounds[1:])):
            l -= mz_overlap / 2 + l * ppm * 1e-6
            r += mz_overlap / 2 + r * ppm * 1e-6
            mz_segments.append((l, r))
        return mz_segments

    sp_n = len(imzml_parser.coordinates)
    sample_sp_gen = ds_sample_gen(imzml_parser, sp_n, sample_ratio)
    min_mzs, max_mzs, n_mzs = zip(*((mzs[0], mzs[-1], len(mzs)) for (sp_id, mzs, ints) in sample_sp_gen))
    min_mz, max_mz, n_mz = min(min_mzs), max(max_mzs), sum(n_mzs)

    total_n_mz = n_mz * 1 / sample_ratio
    segm_n = optimal_segm_n(total_n_mz, mz_per_segm=5e5, min_i=4, max_i=8)

    centr_mzs = centroids_df[(centroids_df.mz > min_mz) & (centroids_df.mz < max_mz)].mz.values

    segm_bounds_q = [i * 1 / segm_n for i in range(0, segm_n + 1)]
    segm_bounds = [np.quantile(centr_mzs, q) for q in segm_bounds_q]

    segments = bounds_to_segments(segm_bounds)
    logger.info(f'Generated {len(segments)} m/z segments: {segments[0]}...{segments[-1]}')
    return np.array(segments)


def segment_centroids(centr_df, mz_segments, centr_segm_path):
    logger.info(f'Segmenting centroids into {len(mz_segments)} segments')

    formula_segments = {}
    for segm_i in range(len(mz_segments))[::-1]:
        logger.debug(f'Segment {segm_i}')

        segm_min_mz, segm_max_mz = mz_segments[segm_i]

        segm_df = centr_df[(~centr_df.formula_i.isin(formula_segments))
                           & (centr_df.mz > segm_min_mz)
                           & (centr_df.mz < segm_max_mz)]

        by_fi = segm_df.groupby('formula_i').peak_i
        formula_min_peak = by_fi.min()
        formula_max_peak = by_fi.max()

        formula_inds = set(formula_min_peak[formula_min_peak == 0].index)
        formula_inds &= set(formula_max_peak[formula_max_peak > 0].index)

        for f_i in formula_inds:
            formula_segments[f_i] = segm_i

    rmtree(centr_segm_path, ignore_errors=True)
    centr_segm_path.mkdir(parents=True)

    logger.info(f'Saving segments to {centr_segm_path}')
    centr_segm_df = centr_df.join(pd.Series(formula_segments, name='segm'), on='formula_i', how='inner')
    for segm_i, df in centr_segm_df.groupby('segm'):
        df.to_msgpack(f'{centr_segm_path}/{segm_i}')


def ds_dims(coordinates):
    min_x, min_y = np.amin(coordinates, axis=0)
    max_x, max_y = np.amax(coordinates, axis=0)
    nrows, ncols = max_y - min_y + 1, max_x - min_x + 1
    return nrows, ncols


def determine_spectra_order(coordinates):
    _coord = np.array(coordinates)
    _coord = np.around(_coord, 5)  # correct for numerical precision
    _coord -= np.amin(_coord, axis=0)

    _, ncols = ds_dims(coordinates)
    pixel_indices = _coord[:, 1] * ncols + _coord[:, 0]
    pixel_indices = pixel_indices.astype(np.int32)
    return pixel_indices


def segment_spectra(imzml_parser, coordinates, mz_segments, ds_segments_path):

    def chunk_list(l, size=5000):
        n = (len(l) - 1) // size + 1
        for i in range(n):
            yield l[size * i:size * (i + 1)]

    def segment_spectra_chunk(sp_inds, mzs, ints):
        for segm_i, (l, r) in enumerate(mz_segments):
            mask = (mzs > l) & (mzs < r)
            n = mask.sum()
            a = np.zeros((n, 3))
            a[:, 0] = sp_inds[mask]
            a[:, 1] = mzs[mask]
            a[:, 2] = ints[mask]
            (pd.DataFrame(a, columns=['idx', 'mz', 'int'])
             .to_msgpack(ds_segments_path / f'{segm_i}', append=True))

    logger.info(f'Segmenting spectra into {len(mz_segments)} segments')

    rmtree(ds_segments_path, ignore_errors=True)
    ds_segments_path.mkdir(parents=True)

    sp_indices = determine_spectra_order(coordinates)

    chunk_size = 5000
    coord_chunk_it = chunk_list(coordinates, chunk_size)

    sp_i = 0
    sp_inds, mzs, ints = [], [], []
    for ch_i, coord_chunk in enumerate(coord_chunk_it):
        logger.debug(f'Segmenting chunk {ch_i}')

        for x, y in coord_chunk:
            mzs_, ints_ = map(np.array, imzml_parser.getspectrum(sp_i))
            sp_idx = sp_indices[sp_i]
            sp_inds.append(np.ones_like(mzs_) * sp_idx)
            mzs.append(mzs_)
            ints.append(ints_)
            sp_i += 1

        segment_spectra_chunk(np.concatenate(sp_inds),
                              np.concatenate(mzs),
                              np.concatenate(ints))
        sp_inds, mzs, ints = [], [], []


def make_sample_area_mask(coordinates):
    sp_indices = determine_spectra_order(coordinates)
    nrows, ncols = ds_dims(coordinates)
    sample_area_mask = np.zeros(ncols * nrows, dtype=bool)
    sample_area_mask[sp_indices] = True
    return sample_area_mask.reshape(nrows, ncols)


def create_process_segment(ds_segments_path, centr_segm_path,
                           coordinates, image_gen_conf, target_formula_inds):
    sample_area_mask = make_sample_area_mask(coordinates)
    nrows, ncols = ds_dims(coordinates)
    compute_metrics = make_compute_image_metrics(sample_area_mask, nrows, ncols, image_gen_conf)

    def process_segment(segm_i):
        logger.info(f'Processing segment {segm_i}')
        logger.info(f'Reading spectra data from {ds_segments_path / str(segm_i)}')
        data = pd.read_msgpack(ds_segments_path / f'{segm_i}')
        if type(data) == pd.DataFrame:
            sp_df = data
        else:
            sp_df = pd.concat(data)
        logger.info(f'Reading centroids data from {centr_segm_path / str(segm_i)}')
        centr_df = pd.read_msgpack(centr_segm_path / f'{segm_i}')

        formula_images_gen = gen_iso_images(sp_df.idx.values, sp_df.mz.values, sp_df.int.values,
                                            centr_df, nrows, ncols, ppm=3, min_px=1)
        formula_metrics_df, formula_images = \
            formula_image_metrics(formula_images_gen, compute_metrics, target_formula_inds)
        logger.info(f'Segment {segm_i} finished')
        return formula_metrics_df, formula_images

    return process_segment
