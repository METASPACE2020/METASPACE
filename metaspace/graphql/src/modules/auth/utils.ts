export const DbSchemaName = 'auth';

export const createExpiry = (minutes: number=10) => {
  const now = new Date().valueOf();
  return new Date( now + minutes * 60 * 1000);
};