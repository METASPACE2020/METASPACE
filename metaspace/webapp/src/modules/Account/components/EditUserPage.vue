<template>
  <div class="main-content">
    <div class="user-edit-page">
      <transfer-datasets-dialog
        v-if="showTransferDatasetsDialog"
        :currentUserId="currentUser && currentUser.id"
        :groupName="invitingGroup && invitingGroup.name"
        :isInvited="true"
        @accept="handleAcceptTransferDatasets"
        @close="handleCloseTransferDatasetsDialog"
      />
      <el-dialog
        title="Delete account"
        custom-class="delete-account-dialog"
        :visible.sync="showDeleteAccountDialog"
        width="500px"
        :lock-scroll="false">
        <p>If you delete your account, you will lose access to all private datasets.
        </p>
        <p>Please select whether you would like to delete all your datasets or keep them within METASPACE accessible by the group members:
        </p>
        <!--Changed the Text below to the above version after discussion with Theo, plz consider if it fits-->
        <!--If you delete your account, you will lose access to any datasets, groups and-->
        <!--projects that have been explicitly shared with you-->
        <el-checkbox v-model="delDatasets" style="margin: 10px 20px">
          Delete datasets that I have submitted
        </el-checkbox>
        <div style="margin: 10px 0 20px; text-align: right;">
          <el-button title="Cancel" @click="closeDeleteAccountDialog">Cancel</el-button>
          <el-button
            type="danger"
            title="Delete account"
            @click="deleteAccount()"
            :loading="isUserDeletionLoading">
            Delete account</el-button>
        </div>
        <p>
          <b>Note:</b> if you choose not to delete the datasets now, you will still be able to have them
          deleted later by emailing the <a href="mailto:contact@metaspace2020.eu">METASPACE administrators</a>.
        </p>
      </el-dialog>
      <el-row id="edit-user-page">
        <el-row>
          <el-col :span="16">
            <h2>User details</h2>
          </el-col>
          <el-col :span="8">
            <el-button title="Save"
                       type="primary"
                       @click="updateUserDetails"
                       class="saveButton"
                       :loading="isUserDetailsLoading">
              Save
            </el-button>
          </el-col>
        </el-row>
        <el-row :gutter="20">
          <el-form :disabled="isUserDetailsLoading" :rules="rules" :model="model" ref="form">
            <div class="user-details" style="padding-left: 15px;">
              <el-col :span="12">
                <div class="fullname">
                  <el-form-item prop="name" label="Full name:">
                    <el-input v-model="model.name" />
                  </el-form-item>
                </div>
              </el-col>
              <el-col :span="12">
                <div class="email">
                  <el-form-item prop="email" label="E-mail:">
                    <el-input v-model="model.email" />
                  </el-form-item>
                </div>
              </el-col>
            </div>
          </el-form>
        </el-row>

        <div class="groups">
          <h2>Groups</h2>
          <el-table
            :data="groupsData"
            style="width: 100%;padding-left: 15px;">
            <el-table-column
              prop="name"
              label="Group"
              width="180">
            </el-table-column>
            <el-table-column
              prop="roleName"
              label="Role"
              width="280">
            </el-table-column>
            <el-table-column
              prop="numDatasets"
              label="Dataset contributed">
            </el-table-column>
            <el-table-column>
              <template slot-scope="scope">
                <el-button
                  v-if="scope.row.role === 'MEMBER'"
                  size="mini"
                  icon="el-icon-arrow-right"
                  @click="leaveGroup('leave', scope.row)">
                  Leave
                </el-button><el-button
                  v-if="scope.row.role === 'PRINCIPAL INVESTIGATOR'"
                  size="mini"
                  icon="el-icon-arrow-right"
                  disabled>
                  Leave
                </el-button>
                <el-button
                  v-if="scope.row.role === 'INVITED'"
                  size="mini"
                  type="success"
                  @click="acceptInvitation(scope.row)"
                  icon="el-icon-check">
                  Accept
                </el-button>
                <el-button
                  v-if="scope.row.role === 'INVITED'"
                  size="mini"
                  icon="el-icon-close"
                  @click="leaveGroup('leaveGroup', scope.row)">
                  Decline
                </el-button>
              </template>
            </el-table-column>
          </el-table>
          <p>Primary group:</p>
          <el-select v-model="primaryGroupId" placeholder="Select"
                     style="padding-left: 15px;">
            <el-option
              v-for="item in groupsData"
              :key="item.id"
              :label="item.name"
              :value="item.id">
            </el-option>
          </el-select>
        </div>
        <!--The section below will be introduced in vFuture-->
        <!--<div class="notifications" style="margin-top: 30px">-->
          <!--<h2>Notifications</h2>-->
          <!--<div class="notification-list" style="padding-left: 10px">-->
            <!--<el-row :gutter="0">-->
              <!--<el-col :span="12"><p>Send an email when:</p>-->
                <!--<el-checkbox-group-->
                  <!--class="notifications">-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox-->
                      <!--label="My dataset's processing has successfully finished"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="My dataset's processing failed"></el-checkbox>-->
                  <!--</div>-->
                <!--</el-checkbox-group>-->
                <!--<p>Show a notification when:</p>-->
                <!--<el-checkbox-group :indeterminate="isIndeterminate" v-model="checkList"-->
                                   <!--class="notifications">-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="My dataset's processing has successfully finished"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="A dataset is added to a group or project that I belong to"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="A public dataset has been added to the queue"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="A public dataset has finished processing"></el-checkbox>-->
                  <!--</div>-->
                <!--</el-checkbox-group>-->
              <!--</el-col>-->
            <!--</el-row>-->
          <!--</div>-->
        <!--</div>-->
        <div class="delete-account" style="margin-top: 45px">
          <h2>Delete account</h2>
          <p style="width: 100%;padding-left: 15px;">
            If you delete your METASPACE account, you can either delete all your datasets or keep them within METASPACE.
            For the latter, the private data will still be accessible by the group members only.
          </p>
        </div>
        <el-row>
          <el-button
            type="danger"
            title="Delete account"
            @click="openDeleteAccountDialog()"
            style="float:right; margin-top:15px">
            Delete account
          </el-button>
        </el-row>
      </el-row>
    </div>
  </div>
</template>

<script lang="ts">
  import Vue from 'vue'
  import { Component, Watch } from 'vue-property-decorator'
  import { updateUserMutation, deleteUserMutation, userProfileQuery, UserProfileQuery } from '../../../api/user';
  import { leaveGroupMutation, acceptGroupInvitationMutation } from '../../../api/group';
  import reportError from "../../../lib/reportError";
  import {refreshLoginStatus} from '../../../graphqlClient';
  import {ElForm} from "element-ui/types/form";
  import TransferDatasetsDialog from '../../GroupProfile/TransferDatasetsDialog.vue'
  import emailRegex from '../../../lib/emailRegex';
  import ConfirmAsync from '../../../components/ConfirmAsync';
  import { getRoleName, UserGroupRole } from '../../../api/group';

  interface Model {
    name: string;
    email: string | null;
  }

  interface GroupRow {
    id: string;
    name: string;
    role: UserGroupRole;
    roleName: string;
    numDatasets: number;
  }

  @Component({
    components: {
      TransferDatasetsDialog
    },
    apollo: {
      currentUser: {
        query: userProfileQuery
      }
    }
  })

  export default class EditUserPage extends Vue {
    showDeleteAccountDialog: boolean = false;
    isUserDetailsLoading: boolean = false;
    isUserDeletionLoading: boolean = false;
    showTransferDatasetsDialog: boolean = false;

    currentUser: UserProfileQuery | null = null;
    model: Model = {
      name: '',
      email: ''
    };
    primaryGroupId: string | null = null;
    invitingGroup: GroupRow | null = null;

    delDatasets: boolean = false;
    rules: object = {
      name: [
        {required: true, min: 3, max: 50 , message: 'Please enter a correct fullname', trigger: "blur"}
      ],
      email: [
        {required: true, pattern: emailRegex,
        message: 'Please enter a correct Email address', trigger: "blur"
      }]
    };

    @Watch('currentUser', {deep: true})
    onCurrentUserChanged(this: any) {
      this.model.name = this.currentUser.name;
      this.model.email = this.currentUser.email;
      this.primaryGroupId = this.currentUser.primaryGroup ? this.currentUser.primaryGroup.group.id : null;
    }

    openDeleteAccountDialog() {
      this.showDeleteAccountDialog = true;
    }

    closeDeleteAccountDialog() {
      this.showDeleteAccountDialog = false;
    }

    get groupsData(): GroupRow[] {
      if (this.currentUser == null || this.currentUser.groups == null) {
        return [];
      }
      return this.currentUser.groups.map(({group, numDatasets, role}) => {
        const {id, name} = group;
        const roleName = getRoleName(role);
        return {id, name, role, roleName, numDatasets}
      });
    }

    async updateUserDetails() {
      try {
        await (this.$refs.form as ElForm).validate();
      } catch (err) {
        return;
      }
      try {
        if (this.currentUser!.email !== this.model.email) {
          try {
            await this.$confirm(
              "Are you sure you want to change email address? A verification email will be sent to your new address to confirm the change.",
              "Confirm email address change", {
                confirmButtonText: "Yes, send verification email",
                lockScroll: false
              });
          } catch {
            return
          }
        }
        this.isUserDetailsLoading = true;
        await this.$apollo.mutate({
          mutation: updateUserMutation,
          variables: {
            update: {
              id: this.currentUser!.id,
              name: this.model.name,
              email: this.model.email,
              primaryGroupId: this.primaryGroupId
            }
          },
        });
        await this.$apollo.queries.currentUser.refetch();
          this.$message({
            type: "success",
            message: "A verification link has been sent to your new email address. " +
            "Please click the link in this email to confirm the change."
          });
      } catch(err) {
        reportError(err);
      } finally {
        this.isUserDetailsLoading = false;
      }
    }

    @ConfirmAsync(function(action: string, groupRow: GroupRow) {
      return {
        message: (action === 'leave') ? `Are you sure you want to leave ${groupRow.name}?` :
          `Are you sure you want to decline the invitation to ${groupRow.name}?`,
        confirmButtonText: (action === 'leave') ? "Yes, leave the group" :
          "Yes, decline the invitation",
        confirmButtonLoadingText: 'Leaving...'
      };
    })
    async leaveGroup(action: string, groupRow: GroupRow) {
      await this.$apollo.mutate({
        mutation: leaveGroupMutation,
        variables: {
          groupId: groupRow.id
        }
      });
      this.$message({
        message: "You have declined the invitation"
      });
      await this.$apollo.queries.currentUser.refetch();
      if (action === 'leave') {
        this.$message({
          message: "You have successfully left the group"
        });
      }
    }

    async deleteAccount(this: any) {
      try {
        this.isUserDeletionLoading = true;
        await this.$apollo.mutate({
          mutation: deleteUserMutation,
          variables: {
            id: this.currentUser.id,
            deleteDatasets: this.delDatasets
          }
        });
        this.$message({
          type: "success",
          message: "You have successfully deleted your account!"
        })
      } catch(err) {
        reportError(err);
      } finally {
        await refreshLoginStatus();
        this.closeDeleteAccountDialog();
        this.isUserDeletionLoading = false;
      }
    }

    async acceptInvitation(groupRow: GroupRow) {
      this.showTransferDatasetsDialog = true;
      this.invitingGroup = groupRow;
    }

    async handleAcceptTransferDatasets(selectedDatasetIds: string[]) {
      try {
        await this.$apollo.mutate({
          mutation: acceptGroupInvitationMutation,
          variables: { groupId: this.invitingGroup!.id, bringDatasets: selectedDatasetIds },
        });
        await this.$apollo.queries.currentUser.refetch();
        this.$message({
          type: "success",
          message: "You have successfully joined the group!"
        });
        this.showTransferDatasetsDialog = false;
      } catch(err) {
        reportError(err);
      } finally {
        this.showTransferDatasetsDialog = false;
      }
    }

    handleCloseTransferDatasetsDialog() {
      this.showTransferDatasetsDialog = false;
    }
  }
</script>

<style scoped>
  .main-content {
    padding: 100px 20px 20px 20px;
    display: flex;
    justify-content: center;
  }

  .user-edit-page {
    max-width: 950px;
  }

  .saveButton {
    width: 100px;
    padding: 8px;
    float: right;
    margin: 20px 0;
  }

  .user-details {
    display: inline-block;
    width: 600px;
  }

  /deep/ .delete-account-dialog .el-dialog__body {
    padding: 0 20px 20px 20px;
  }

  /* Uncomment when the vFuture notifications will be introduced
  /*.notifications_checkbox {*/
    /*margin-left: 0;*/
    /*padding: 0;*/
  /*}*/
</style>
