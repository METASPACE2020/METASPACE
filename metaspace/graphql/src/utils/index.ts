import {logger as _logger} from '../../utils';
import {LoggerInstance} from 'winston';

export {createConnection, findUserByEmail} from './db';
export {smAPIRequest} from './smAPI';

export const logger = _logger as LoggerInstance;

export type LooselyCompatible<T> =
  {[K in keyof T]?: T[K] extends (string | number | null | undefined) ? T[K] | null : any};
