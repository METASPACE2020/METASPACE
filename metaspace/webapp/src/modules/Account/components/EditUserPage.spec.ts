import { mount, Wrapper } from '@vue/test-utils';
import ElementUI from 'element-ui';
import Vue from 'vue';
import registerMockComponent from '../../../../tests/utils/registerMockComponent';
import { restoreConsole, suppressConsoleWarn } from '../../../../tests/utils/suppressConsole';
import EditUserPage from './EditUserPage.vue';

Vue.use(ElementUI);
registerMockComponent('el-dialog'); // ElDialogs mount their content somewhere else in the DOM. Mock it out so that the snapshot includes the content.

describe('EditUserPage', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    suppressConsoleWarn('async-validator:');
  });

  afterEach(async () => {
    jest.useRealTimers();
    restoreConsole();
  });


  it('should match snapshot', async () => {
    const wrapper = mount(EditUserPage, { sync: false });
    wrapper.setData({
      isLoaded: true,
      currentUser: {
        id: '22333',
        name: 'foo',
        role: 'user'
      },
      model: {
        name: 'foo',
        email: 'bar'
      }
    });
    await Vue.nextTick();
    jest.runAllTimers();

    expect(wrapper).toMatchSnapshot();
  });

});
