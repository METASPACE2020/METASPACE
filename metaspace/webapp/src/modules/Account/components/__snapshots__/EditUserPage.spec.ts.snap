// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserPage should match snapshot 1`] = `
<div class="main-content">
  <!---->
  <mock-el-dialog title="Delete account" custom-class="delete-account-dialog" width="500px">
    <p>If you delete your account, you will lose access to all private datasets.
    </p>
    <p>Please select whether you would like to delete all your datasets or keep them within METASPACE accessible by the group members:
    </p>
    <label role="checkbox" class="el-checkbox" style="margin: 10px 20px;"><span aria-checked="mixed" class="el-checkbox__input"><span class="el-checkbox__inner"></span>
      <input type="checkbox" aria-hidden="true" class="el-checkbox__original" value="">
      </span><span class="el-checkbox__label">
      Delete datasets that I have submitted
    <!----></span></label>
    <div style="margin: 10px 0px 20px; text-align: right;">
      <button type="button" class="el-button el-button--default" title="Cancel">
        <!---->
        <!----><span>Cancel</span></button>
      <button type="button" class="el-button el-button--danger" title="Delete account">
        <!---->
        <!----><span>
        Delete account</span></button>
    </div>
    <p>
      <b>Note:</b> if you choose not to delete the datasets now, you will still be able to have them deleted later by emailing the
      <a href="mailto:contact@metaspace2020.eu">METASPACE administrators</a>.
    </p>
  </mock-el-dialog>
  <div class="user-edit-page">
    <div class="el-row">
      <div class="el-col el-col-16">
        <h2>User details</h2>
      </div>
      <div class="el-col el-col-8">
        <button type="button" class="el-button saveButton el-button--primary" title="Save">
          <!---->
          <!----><span>
          Save
        </span></button>
      </div>
    </div>
    <div class="el-row" style="margin-left: -10px; margin-right: -10px;">
      <form class="el-form">
        <div class="user-details" style="padding-left: 15px;">
          <div class="el-col el-col-12" style="padding-left: 10px; padding-right: 10px;">
            <div>
              <div class="el-form-item is-required">
                <label for="name" class="el-form-item__label">Full name:</label>
                <div class="el-form-item__content">
                  <div class="el-input">
                    <!---->
                    <input type="text" autocomplete="off" class="el-input__inner">
                    <!---->
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
              </div>
            </div>
          </div>
          <div class="el-col el-col-12" style="padding-left: 10px; padding-right: 10px;">
            <div>
              <div class="el-form-item is-required">
                <label for="email" class="el-form-item__label">E-mail:</label>
                <div class="el-form-item__content">
                  <div class="el-input">
                    <!---->
                    <input type="text" autocomplete="off" class="el-input__inner">
                    <!---->
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div>
      <h2>Groups</h2>
      <div class="el-table el-table--fit el-table--scrollable-x el-table--enable-row-hover" style="width: 100%; padding-left: 15px;">
        <div class="hidden-columns">
          <div></div>
          <div></div>
          <div></div>
          <div></div>
        </div>
        <div class="el-table__header-wrapper">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 620px;">
            <colgroup>
              <col name="el-table_1_column_1" width="180">
              <col name="el-table_1_column_2" width="280">
              <col name="el-table_1_column_3" width="80">
              <col name="el-table_1_column_4" width="80">
            </colgroup>
            <thead class="">
              <tr class="">
                <th colspan="1" rowspan="1" class="el-table_1_column_1     is-leaf">
                  <div class="cell">Group</div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_1_column_2     is-leaf">
                  <div class="cell">Role</div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_1_column_3     is-leaf">
                  <div class="cell">Dataset contributed</div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_1_column_4     is-leaf">
                  <div class="cell"></div>
                </th>
              </tr>
            </thead>
          </table>
        </div>
        <div class="el-table__body-wrapper is-scrolling-left">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 620px;">
            <colgroup>
              <col name="el-table_1_column_1" width="180">
              <col name="el-table_1_column_2" width="280">
              <col name="el-table_1_column_3" width="80">
              <col name="el-table_1_column_4" width="80">
            </colgroup>
            <tbody>
              <!---->
            </tbody>
          </table>
          <div class="el-table__empty-block" style="width: 620px;"><span class="el-table__empty-text">暂无数据</span></div>
          <!---->
        </div>
        <!---->
        <!---->
        <!---->
        <!---->
        <div class="el-table__column-resize-proxy" style="display: none;"></div>
      </div>
      <!---->
    </div>
    <div style="margin-top: 45px;">
      <h2>Delete account</h2>
      <p style="width: 100%; padding-left: 15px;">
        If you delete your METASPACE account, you can either delete all your datasets or keep them within METASPACE. For the latter, the private data will still be accessible by the group members only.
      </p>
    </div>
    <div class="el-row">
      <button type="button" class="el-button el-button--danger" title="Delete account" style="float: right; margin-top: 15px;">
        <!---->
        <!----><span>
        Delete account
      </span></button>
    </div>
    <div class="el-loading-mask" style="display: none;">
      <div class="el-loading-spinner">
        <svg viewBox="25 25 50 50" class="circular">
          <circle cx="50" cy="50" r="20" fill="none" class="path"></circle>
        </svg>
        <!---->
      </div>
    </div>
  </div>
</div>
`;
